jscomp/others/js_json_decode.mli

```diff
@@ -0,0 +1,291 @@
+(* Copyright (C) 2015-2016 Bloomberg Finance L.P.
+ * 
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Lesser General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * In addition to the permissions granted to you by the LGPL, you may combine
+ * or link a "work that uses the Library" with a publicly distributed version
+ * of this file to produce a combined library or application, then distribute
+ * that combined work under the terms of your choosing, with no requirement
+ * to comply with the obligations normally placed on you by section 4 of the
+ * LGPL version 3 (or the corresponding section of a later version of the LGPL
+ * should you choose to use a later version).
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Lesser General Public License for more details.
+ * 
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA 02111-1307, USA. *)
+
+
+(** Provides a set of low level combinator primitives to decode Js.Json.t data
+structures
+
+A decoder combinator will return an [('a, string) result] where [Ok of 'a]
+indicates that the decoding succeeded and should contain the decoded value.
+[Error of string] indicates the decoding failed and should contain a [string]
+message explaining why.
+
+Decoders are designed to be combined to produce more complex decoders that can
+decode arbitrary data structures, though the emphasis for this library is for
+it to be {i possible} to decode any given data structure, not necessarily for
+it to be {i convenient}. For convenience you should look towards opinionated
+third-party libraries.
+*)
+
+type 'a decoder = Js.json -> ('a, string) Result.result
+(** The type of a decoder combinator *)
+
+val boolean : Js.boolean decoder
```

Comment:
I am not sure this is the right API.
I am more in favor of raise when decoding fails, the reason is that when decoding, you don't want to do error checking everytime which is expensive, why Elm is in favor of such API partly is because it does not support exception
cc @mransan